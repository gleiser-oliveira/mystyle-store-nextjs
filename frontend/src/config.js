// Client side config constants
export const DOCKER_BACKEND_URL = `http://backend:4000`;
export const BROWSER_TO_BACKEND_URL = `http://localhost:4000`;
export const perPage = 4;