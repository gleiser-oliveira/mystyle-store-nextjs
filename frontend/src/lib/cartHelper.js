export const numberOfProducts = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.quantity), 0);
}

export const calcTotalCartPrice = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.product.price * b.quantity), 0);
}

export const formatPrice = (priceInCents) => {
    const options = {
        style: 'currency',
        currency: 'USD',
        minimumFractionDigits: 2,
    };
    // if its a whole, dollar amount, leave off the .00
    if (priceInCents % 100 === 0)
        options.minimumFractionDigits = 0;

    const formatter = new Intl.NumberFormat('en-US', options);
    return formatter.format(priceInCents / 100);
}