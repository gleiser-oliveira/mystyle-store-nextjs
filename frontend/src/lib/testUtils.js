import casual from 'casual';
import { act } from 'react-dom/test-utils';
import { CURRENT_USER_QUERY } from '../components/authentication/AuthenticationOperationsGQL';

// seeded for constant results
casual.seed(123);

const fakeUser = overrides => ({
    __typename: 'User',
    id: '4234',
    name: casual.name,
    email: casual.email,
    permissions: ['USER', 'PRODUCT_CREATE'],
    cart: [],
    ...overrides,
});

const fakeUser2 = () => ({
    __typename: 'User',
    id: '2222',
    name: casual.name,
    email: casual.email,
    permissions: ['USER'],
    orders: [],
    cart: [],
});

const fakeAdminUser = () => ({
    __typename: 'User',
    id: '1111',
    name: casual.name,
    email: casual.email,
    permissions: ['ADMIN', 'PRODUCT_CREATE'],
    orders: [],
    cart: [],
});

const fakeProduct = () => ({
    __typename: 'Product',
    id: 'QWERTY0123',
    title: 'A Fake Product',
    price: 5000,
    description: 'This is a fake product for testing.',
    image: 'product.jpg',
    largeImage: 'largeProduct.jpg',
    user: fakeUser(),
});

const fakeProduct2 = () => ({
    __typename: 'Product',
    id: 'AS012',
    title: 'Another Fake Product',
    price: 2999,
    description: 'This is another fake product for testing.',
    image: 'product2.jpg',
    largeImage: 'bigProduct2.jpg',
    user: fakeUser(),
});

const fakeOrderProduct = () => ({
    __typename: 'OrderProduct',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
});

const fakeOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    products: [fakeOrderProduct(), fakeOrderProduct()],
    createdAt: 'May 1, 2020 00:00:00',
    user: fakeUser(),
});


const fakeCartProduct = overrides => ({
    __typename: 'CartProduct',
    id: 'prod123',
    quantity: 2,
    product: fakeProduct(),
    user: fakeUser(),
    ...overrides,
});

const fakeCartProduct2 = overrides => ({
    __typename: 'CartProduct',
    id: 'prod312',
    quantity: 2,
    product: fakeProduct2(),
    user: fakeUser(),
    ...overrides,
});

const cartUser = fakeUser();
const fakeCart = () => ([
    fakeCartProduct({user: cartUser}),
    fakeCartProduct2({user: cartUser}),
]);

// Fake LocalStorage
class LocalStorageMock {
    constructor() {
        this.store = {};
    }

    clear() {
        this.store = {};
    }

    getItem(key) {
        return this.store[key] || null;
    }

    setItem(key, value) {
        this.store[key] = value.toString();
    }

    removeItem(key) {
        delete this.store[key];
    }
}

// mocks
const signedInMocks = (user = fakeUser) => { return [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: user(),
            }
        }
    }
]};

const notSignedInMocks = () => ([
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: null,
            }
        }
    }
]);

const updateWrapper = async (wrapper, ms = 0) => {
    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, ms));
        wrapper.update();
    });
}

export {
    LocalStorageMock,
    fakeProduct,
    fakeProduct2,
    fakeUser,
    fakeUser2,
    fakeAdminUser,
    fakeCartProduct,
    fakeCartProduct2,
    fakeOrder,
    fakeCart,
    fakeOrderProduct,
    signedInMocks,
    notSignedInMocks,
    updateWrapper,
};
