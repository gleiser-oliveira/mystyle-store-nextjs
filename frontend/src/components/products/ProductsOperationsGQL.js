import gql from 'graphql-tag';
import { perPage } from '../../config';

export const ALL_PRODUCTS_QUERY = gql`
    query ALL_PRODUCTS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
        products(first: $first, skip: $skip, orderBy: createdAt_DESC) {
            id
            title
            price
            description
            image
            largeImage
            user {
                id
            }
        }
    }
`;

export const SINGLE_PRODUCT_QUERY = gql`
    query SINGLE_PRODUCT_QUERY($id: ID!) {
        product(where: { id: $id }) {
            id
            title
            description
            price
            image
            largeImage
            user {
                id
            }
        }
    }
`;

export const CREATE_PRODUCT_MUTATION = gql`
    mutation CREATE_PRODUCT_MUTATION (
        $title: String!
        $description: String!
        $price: Int!
        $image: String
        $largeImage: String
    ) {
        createProduct (
            title: $title 
            description: $description 
            price: $price
            image: $image
            largeImage: $largeImage
        ) {
            id
        }
    }
`;

export const UPDATE_PRODUCT_MUTATION = gql`
    mutation UPDATE_PRODUCT_MUTATION (
        $id: ID!
        $title: String
        $description: String
        $price: Int
    ) {
        updateProduct (
            id: $id
            title: $title 
            description: $description 
            price: $price
        ) {
            id
        }
    }
`;

export const DELETE_PRODUCT_MUTATION = gql`
    mutation DELETE_PRODUCT_MUTATION($id: ID!) {
        deleteProduct(id: $id) {
            id
        }
    }
`;

export const PRODUCTS_PAGINATION_QUERY = gql`
  query PRODUCTS_PAGINATION_QUERY {
    productsConnection {
      aggregate {
        count
      }
    }
  }
`;
