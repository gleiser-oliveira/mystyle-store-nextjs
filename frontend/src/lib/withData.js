import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from './ClientGQL/LocalGQL';
import { DOCKER_BACKEND_URL, BROWSER_TO_BACKEND_URL } from '../config';

function createClient({ headers }) {
    // if using docker, the URI to the Yoga server must be either:
    // DOCKER_BACKEND_URL = the name of the backend server set in the docker network (used in SSR fetches)
    // BROWSER_TO_BACKEND_URL = the URL of the backend server that can be accessed from the client (used in the client's browser fetches)
    const isInsideFrontendServer = typeof window === 'undefined';
    const uri = isInsideFrontendServer ? DOCKER_BACKEND_URL : BROWSER_TO_BACKEND_URL;
    return new ApolloClient({
        uri: uri,
        request: operation => {
            operation.setContext({
                fetchOptions: {
                    credentials: 'include',
                },
                headers,
            });
        },
        // local data
        clientState: {
            resolvers: {
                Mutation: {
                    toggleCart(_, variables, client) {
                        const { cartOpen } = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        // toggle the state
                        const data = {
                            data: { cartOpen: !cartOpen },
                        };
                        client.cache.writeData(data);
                        return data;
                    },
                },
                Query: {
                    localState(_, variables, client) {
                        const localState = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        return localState;
                    }
                }
            },
            defaults: {
                cartOpen: false,
            }
        }
    });
}

// TODO: check getDataFromTree warning - apollo
export default withApollo(createClient);