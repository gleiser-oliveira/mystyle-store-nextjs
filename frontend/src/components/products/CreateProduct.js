import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import { CREATE_PRODUCT_MUTATION, ALL_PRODUCTS_QUERY } from './ProductsOperationsGQL';
import { Form } from '../theme/generalStyles';
import { ErrorMessage } from '../message/Message';

export default class CreateProduct extends Component {
    state = {
        title: '',
        description: '',
        price: 0,
        image: '',
        largeImage: ''
    };

    handleChange = (e) => {
        const { name, type, value } = e.target;
        const val = type === 'number' ? parseFloat(value) : value;
        this.setState({ [name]: val });
    };

    handleSubmit = async(e, createProduct) => {
        e.preventDefault();
        const res = await createProduct();
        Router.push({
            pathname: '/product',
            query: { id: res.data.createProduct.id }
        });
    };

    uploadImage = async(e) => {
        const files = e.target.files;
        const data = new FormData();
        data.append('file', files[0]);
        data.append('upload_preset', 'my-style');
        const res = await fetch('https://api.cloudinary.com/v1_1/go-medias/image/upload', {
            method: 'POST',
            body: data
        });
        const file = await res.json();
        this.setState({
            image: file.secure_url,
            largeImage: file.eager[0].secure_url
        });
    };

    render() {
        return (
            <Mutation mutation={CREATE_PRODUCT_MUTATION} variables={this.state} refetchQueries={[{ query: ALL_PRODUCTS_QUERY }]}>
                {(createProduct, payload) => (
                    <Form onSubmit={async e => { this.handleSubmit(e, createProduct); }}>
                        {payload.error ? <ErrorMessage error={payload.error} /> : null}
                        <fieldset disabled={payload.loading} aria-busy={payload.loading}>
                            <label htmlFor='title'>
                                Title
                                <input
                                    type='text'
                                    id='title'
                                    name='title'
                                    placeholder='Title'
                                    required
                                    value={this.state.title}
                                    onChange={this.handleChange}/>
                            </label>
                            <label htmlFor='title'>
                                Price
                                <input
                                    type='number'
                                    id='price'
                                    name='price'
                                    placeholder='Price'
                                    required
                                    value={this.state.price}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor='title'>
                                Description
                                <textarea
                                    type='text'
                                    id='description'
                                    name='description'
                                    placeholder='Enter a Description'
                                    required
                                    value={this.state.description}
                                    onChange={this.handleChange}/>
                            </label>
                            <label htmlFor='file'>
                                Image
                                <input
                                    type='file'
                                    id='file'
                                    name='file'
                                    placeholder='Upload Image'
                                    onChange={this.uploadImage}/>
                                {this.state.image && <img width='200px' src={this.state.image} alt='Upload Preview' />}
                            </label>
                            <button type='submit'>Submit</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }
}
