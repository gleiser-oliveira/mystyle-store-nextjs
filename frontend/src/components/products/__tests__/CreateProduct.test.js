import { mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import CreateProduct from '../CreateProduct';
import { CREATE_PRODUCT_MUTATION, ALL_PRODUCTS_QUERY } from '../ProductsOperationsGQL';
import { perPage } from '../../../config';
import { fakeProduct, updateWrapper } from '../../../lib/testUtils';

const fakeProductImage = 'https://product.com/prod.jpg';

describe('testing the <CreateProduct/> component', () => {
    it('renders', async() => {
        const wrapper = mount(
            <MockedProvider>
                <CreateProduct/>
            </MockedProvider>
        );
        const form = wrapper.find('form');
        expect(toJSON(form)).toMatchSnapshot();
    });
    it('uploads the image when the corresponding input is changed', async() => {
        // making a mock for the global JS fetch API
        global.fetch = jest.fn().mockResolvedValue({
            json: () => ({
                secure_url: fakeProductImage,
                eager: [{ secure_url: fakeProductImage }],
            }),
        });
        const wrapper = mount(
            <MockedProvider>
                <CreateProduct/>
            </MockedProvider>
        );
        const imgInput = wrapper.find('input[type="file"]');
        imgInput.simulate('change', { target: { files: ['fakeImg.jpg'] } });
        await updateWrapper(wrapper);
        const component = wrapper.find('CreateProduct').instance();
        expect(global.fetch).toHaveBeenCalled();
        expect(component.state.image).toEqual(fakeProductImage);
        expect(component.state.largeImage).toEqual(fakeProductImage);
        global.fetch.mockReset();
    });
    it('updates its state when there are input changes', async() => {
        const fakeProd = fakeProduct();
        const wrapper = mount(
            <MockedProvider>
                <CreateProduct/>
            </MockedProvider>
        );

        const title = wrapper.find('#title');
        const description = wrapper.find('#description');
        const price = wrapper.find('#price');

        title.simulate('change', { target: { value: fakeProd.title, name: 'title' } });
        description.simulate('change', { target: { value: fakeProd.description, name: 'description' } });
        price.simulate('change', { target: { value: fakeProd.price, name: 'price' } });

        // await updateWrapper(wrapper);

        const component = wrapper.find('CreateProduct').instance();

        expect(component.state).toMatchObject({
            title: fakeProd.title,
            description: fakeProd.description,
            price: fakeProd.price,
        });
    });
    it('creates a product when the form is submitted', async() => {
        const fakeProd = fakeProduct();
        const mocks = [
            {
                request: {
                    query: CREATE_PRODUCT_MUTATION,
                    variables: {
                        title: fakeProd.title,
                        description: fakeProd.description,
                        image: '',
                        largeImage: '',
                        price: fakeProd.price,
                    },
                },
                result: {
                    data: {
                        createProduct: {
                            __typename: fakeProd.__typename,
                            id: fakeProd.id,
                        },
                    },
                },
            },
            {
                request: {
                    query: ALL_PRODUCTS_QUERY,
                    variables: {
                        skip: 0,
                        first: perPage
                    },
                },
                result: {
                    data: {
                        products: []
                    }
                }
            },
        ];
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <CreateProduct/>
            </MockedProvider>
        );

        const title = wrapper.find('#title');
        const description = wrapper.find('#description');
        const price = wrapper.find('#price');

        title.simulate('change', { target: { value: fakeProd.title, name: 'title' } });
        description.simulate('change', { target: { value: fakeProd.description, name: 'description' } });
        price.simulate('change', { target: { value: fakeProd.price, name: 'price' } });

        Router.router = { push: jest.fn() };
        wrapper.find('form').simulate('submit');
        await updateWrapper(wrapper, 50);
        expect(Router.router.push).toHaveBeenCalled();
    });
});
