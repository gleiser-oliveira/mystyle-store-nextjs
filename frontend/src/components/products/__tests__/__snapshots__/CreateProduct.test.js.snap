// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the <CreateProduct/> component renders 1`] = `
<form
  className="generalStyles__Form-sc-1oz5zu1-1 kCRZkl"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="title"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="title"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a Description"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload Image"
        type="file"
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
